packages/components/src/site-thumbnail/use-mshots-img.tsx

```diff
@@ -27,52 +27,55 @@ export const useMshotsImg = (
 	src: string,
 	options: MShotsOptions
 ): {
-	src: string;
 	isLoading: boolean;
 	isError: boolean;
-	imgRef: React.MutableRefObject< HTMLImageElement | null >;
+	imgProps: Partial< React.ImgHTMLAttributes< HTMLImageElement > >;
 } => {
 	const [ retryCount, setRetryCount ] = useState( 0 );
 	const mshotUrl = useMemo(
 		() => mshotsUrl( src, options, retryCount ),
 		[ src, options, retryCount ]
 	);
-	const imgRef = useRef< HTMLImageElement >( null );
 	const [ isLoading, setIsLoading ] = useState( true );
 	const [ isError, setIsError ] = useState( false );
 
+	const timeout = useRef< ReturnType< typeof setTimeout > >();
+
+	const onLoad = useCallback(
+		( event: React.SyntheticEvent< HTMLImageElement, Event > ) => {
+			if ( ! mshotUrl.length ) {
+				return;
+			}
+
+			// MShot Loading image is 400x300px.
+			// MShot 404 image is 748×561px
+			setIsLoading( true );
+			const hasLoadingImgDimensions =
+				event.currentTarget.naturalWidth === 400 && event.currentTarget.naturalHeight === 300;
+			if ( ! hasLoadingImgDimensions ) {
+				setIsLoading( false );
+			} else if ( retryCount < MAXTRIES ) {
+				// Only refresh 10 times
+				timeout.current = setTimeout(
+					() => setRetryCount( ( retryCount ) => retryCount + 1 ),
+					retryCount * 500
+				);
+			}
+		},
+		[ retryCount, mshotUrl.length ]
+	);
+
+	const onError = useCallback( () => {
+		setIsError( true );
+	}, [] );
+
 	useEffect( () => {
-		let timeout: number;
-		if ( mshotUrl.length > 0 && imgRef?.current && ! imgRef.current.onload ) {
-			imgRef.current.onload = () => {
-				// MShot Loading image is 400x300px.
-				// MShot 404 image is 748×561px
-				setIsLoading( true );
-				const hasLoadingImgDimensions =
-					imgRef?.current?.naturalWidth === 400 && imgRef?.current.naturalHeight === 300;
-				if ( ! hasLoadingImgDimensions ) {
-					setIsLoading( false );
-				} else if ( retryCount < MAXTRIES ) {
-					// Only refresh 10 times
-					timeout = setTimeout(
-						() => setRetryCount( ( retryCount ) => retryCount + 1 ),
-						retryCount * 500
-					);
-				}
-			};
-			imgRef.current.onerror = () => {
-				setIsError( true );
-			};
-		}
-		return () => {
-			clearTimeout( timeout );
-		};
-	}, [ imgRef, mshotUrl, retryCount ] );
+		clearTimeout( timeout.current );
+	}, [ mshotUrl ] );
 
 	return {
-		src: mshotUrl,
 		isLoading,
 		isError,
-		imgRef,
+		imgProps: { onLoad, onError, src: mshotUrl, loading: 'lazy' },
```

Comment:
I like using `imgProps` instead of `imgRef`

Reply:
This is a nice alternative. Removing the onload event in the useEffect cleanup function could also solve this issue ?

Reply:
Sorry I haven't written the PR description yet, but the `onload` handler is what alerted me to the issue. When quickly switching from grid to table view I got a React error in the console about setting state on an unmounted component. The problem is that the `onload` handlers for previous thumbnail components are being called after we've made the switch.

Usually the `onload` handler wouldn't get called once the `<img>` element has been removed from the DOM. So something is holding onto a reference to that element and keeping it alive.

And that seems to be the `imgRef`. I think using it in the `useEffect` dependency list is what does it. Using refs this way (where you have a hook create the ref, and then returning it to be used by someone else) seems rife for memory leaks. So I like the idea of only returning the callbacks. Ownership is less ambiguous; only the `<SiteThumbnail>` component ever has ownership over the `<img>` element.